File: server/routes/modelHubRoutes.ts
import { Router, Request, Response } from "express";
import { fetchCivitaiModelInfo } from "../services/civitaiService";
import { fetchHuggingFaceModelInfo } from "../services/huggingfaceService";
import { 
  executeUnifiedRemoteDownload, 
  COMFYUI_MODEL_CATEGORIES 
} from "../services/modelHubService";

const router = Router();

/**
 * GET /api/model-hub/categories
 * Returns standard ComfyUI model category presets
 */
router.get("/categories", (req: Request, res: Response) => {
  res.json({
    success: true,
    categories: COMFYUI_MODEL_CATEGORIES
  });
});

/**
 * GET /api/model-hub/hf-info
 * Look up Hugging Face repository/file or direct model URL
 */
router.get("/hf-info", async (req: Request, res: Response) => {
  const query = (
    req.query.url ||
    req.query.query ||
    req.query.model ||
    ""
  ).toString().trim();

  const token = (req.query.token as string) || undefined;

  if (!query) {
    return res.status(400).json({
      success: false,
      error: "Missing required parameter 'url' or 'query' (Hugging Face URL, repo ID, or direct model link)."
    });
  }

  try {
    const metadata = await fetchHuggingFaceModelInfo(query, token);
    return res.json({
      success: true,
      data: metadata,
      ...metadata
    });
  } catch (err: any) {
    console.error("[HF Model Info Error]:", err);
    return res.status(400).json({
      success: false,
      error: err.message || "Failed to inspect model on Hugging Face."
    });
  }
});

/**
 * GET /api/model-hub/civitai-info
 * Look up Civitai model metadata
 */
router.get("/civitai-info", async (req: Request, res: Response) => {
  const query = (
    req.query.query ||
    req.query.url ||
    req.query.modelId ||
    req.query.versionId ||
    ""
  ).toString().trim();

  const token = (req.query.token as string) || undefined;

  if (!query) {
    return res.status(400).json({
      success: false,
      error: "Missing required parameter 'query' or 'url' (Civitai Model ID, Version ID, or URL)."
    });
  }

  try {
    const metadata = await fetchCivitaiModelInfo(query, token);
    return res.json({
      success: true,
      data: metadata,
      ...metadata
    });
  } catch (err: any) {
    console.error("[Civitai Model Info Error]:", err);
    return res.status(400).json({
      success: false,
      error: err.message || "Failed to inspect model on Civitai."
    });
  }
});

/**
 * POST /api/model-hub/download-remote
 * Download any model directly to remote GPU ComfyUI instance via SSH
 */
router.post("/download-remote", async (req: Request, res: Response) => {
  const {
    download_url,
    destination_folder,
    filename,
    auth_type,
    api_token,
    civitai_token,
    hf_token,
    remote_host,
    ssh_port = 22,
    ssh_username = "root",
    ssh_password,
    ssh_private_key,
    ssh_key_path,
    remote_comfyui_root
  } = req.body || {};

  if (!download_url) {
    return res.status(400).json({ success: false, error: "Parameter 'download_url' is required." });
  }
  if (!destination_folder) {
    return res.status(400).json({ success: false, error: "Parameter 'destination_folder' is required." });
  }
  if (!filename) {
    return res.status(400).json({ success: false, error: "Parameter 'filename' is required." });
  }
  if (!remote_host) {
    return res.status(400).json({ success: false, error: "Remote Host IP / Address is required for SSH download." });
  }

  const token = api_token || civitai_token || hf_token;

  try {
    const result = await executeUnifiedRemoteDownload({
      download_url,
      destination_folder,
      filename,
      auth_type,
      api_token: token,
      remote_host,
      ssh_port: Number(ssh_port) || 22,
      ssh_username,
      ssh_password,
      ssh_private_key,
      ssh_key_path,
      remote_comfyui_root
    });

    if (!result.success) {
      return res.status(400).json(result);
    }
    return res.json(result);
  } catch (err: any) {
    console.error("[Remote Model Download Error]:", err);
    return res.status(500).json({
      success: false,
      error: err.message || "An unexpected error occurred during remote download."
    });
  }
});

export default router;

File: backend/routes/model_hub_routes.py
from typing import Optional
from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel

from backend.services.civitai_service import fetch_civitai_model_info
from backend.services.huggingface_service import fetch_huggingface_model_info
from backend.services.model_hub_service import (
    COMFYUI_MODEL_CATEGORIES,
    execute_unified_remote_download
)

router = APIRouter(prefix="/model-hub", tags=["Model Ingestion Hub"])

class UnifiedDownloadRequest(BaseModel):
    download_url: str
    destination_folder: str
    filename: str
    auth_type: Optional[str] = None
    api_token: Optional[str] = None
    civitai_token: Optional[str] = None
    hf_token: Optional[str] = None
    remote_host: str
    ssh_port: Optional[int] = 22
    ssh_username: Optional[str] = "root"
    ssh_password: Optional[str] = None
    ssh_private_key: Optional[str] = None
    ssh_key_path: Optional[str] = None
    remote_comfyui_root: Optional[str] = "/workspace/runpod-slim/ComfyUI"

    class Config:
        extra = "allow"

@router.get("/categories")
async def get_model_categories():
    """Returns standard ComfyUI model category presets."""
    return {
        "success": True,
        "categories": COMFYUI_MODEL_CATEGORIES
    }

@router.get("/hf-info")
async def get_huggingface_info(
    url: Optional[str] = Query(None),
    query: Optional[str] = Query(None),
    model: Optional[str] = Query(None),
    token: Optional[str] = Query(None)
):
    """Look up Hugging Face repository/file or direct model URL."""
    search_query = (url or query or model or "").strip()
    if not search_query:
        raise HTTPException(
            status_code=400,
            detail="Missing required parameter 'url' or 'query' (Hugging Face URL, repo ID, or direct model link)."
        )

    try:
        metadata = await fetch_huggingface_model_info(search_query, token_override=token)
        return {
            "success": True,
            "data": metadata,
            **metadata
        }
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))

@router.get("/civitai-info")
async def get_civitai_info(
    query: Optional[str] = Query(None),
    url: Optional[str] = Query(None),
    modelId: Optional[str] = Query(None),
    versionId: Optional[str] = Query(None),
    token: Optional[str] = Query(None)
):
    """Look up Civitai model metadata."""
    search_query = (query or url or modelId or versionId or "").strip()
    if not search_query:
        raise HTTPException(
            status_code=400,
            detail="Missing required parameter 'query' or 'url' (Civitai Model ID, Version ID, or URL)."
        )

    try:
        metadata = await fetch_civitai_model_info(search_query, token_override=token)
        return {
            "success": True,
            "data": metadata,
            **metadata
        }
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))

@router.post("/download-remote")
async def download_remote_model(req: UnifiedDownloadRequest):
    """Download any model directly to remote GPU ComfyUI instance via SSH."""
    if not req.download_url:
        raise HTTPException(status_code=400, detail="Parameter 'download_url' is required.")
    if not req.destination_folder:
        raise HTTPException(status_code=400, detail="Parameter 'destination_folder' is required.")
    if not req.filename:
        raise HTTPException(status_code=400, detail="Parameter 'filename' is required.")
    if not req.remote_host:
        raise HTTPException(status_code=400, detail="Remote Host IP / Address is required for SSH download.")

    token = req.api_token or req.civitai_token or req.hf_token

    result = await execute_unified_remote_download({
        "download_url": req.download_url,
        "destination_folder": req.destination_folder,
        "filename": req.filename,
        "auth_type": req.auth_type,
        "api_token": token,
        "remote_host": req.remote_host,
        "ssh_port": req.ssh_port or 22,
        "ssh_username": req.ssh_username or "root",
        "ssh_password": req.ssh_password,
        "ssh_private_key": req.ssh_private_key,
        "ssh_key_path": req.ssh_key_path,
        "remote_comfyui_root": req.remote_comfyui_root or "/workspace/runpod-slim/ComfyUI"
    })

    if not result.get("success"):
        raise HTTPException(status_code=400, detail=result.get("message") or "Download failed")
    return result
